crate/pkcs11/src/backend.rs

```diff
@@ -0,0 +1,179 @@
+
+use cosmian_kms_client::{ClientConf, KmsClient};
+use native_pkcs11_traits::{
+    Backend, Certificate, DataObject, KeyAlgorithm, PrivateKey, PublicKey, SearchOptions,
+    SignatureAlgorithm,
+};
+use tracing::trace;
+use zeroize::Zeroizing;
+
+use crate::{
+    error::Pkcs11Error,
+    kms_client::{get_kms_client, get_pkcs11_keys},
+};
+
+pub struct CkmsBackend {
+    kms_client: KmsClient,
+}
+
+impl CkmsBackend {
+    /// Instantiate a new CkmsBackend using the `kms.json` file in the local default directory.
+    pub fn instantiate() -> Result<Self, Pkcs11Error> {
+        Ok(CkmsBackend {
+            kms_client: get_kms_client()?,
+        })
+    }
+
+    /// Instantiate a new CkmsBackend using the provided `ClientConf`.
+    #[allow(dead_code)]
+    pub fn instantiate_from_client_conf(client_conf: ClientConf) -> Result<Self, Pkcs11Error> {
+        Ok(CkmsBackend {
+            kms_client: client_conf.initialize_kms_client()?,
+        })
+    }
+}
+
+impl Backend for CkmsBackend {
+    fn name(&self) -> String {
+        trace!("name");
+        "Cosmian KMS".to_string()
+    }
+
+    fn find_certificate(
+        &self,
+        _query: SearchOptions,
+    ) -> native_pkcs11_traits::Result<Option<Arc<dyn Certificate>>> {
+        trace!("find_all_certificates");
+        Ok(None)
+    }
+
+    fn find_all_certificates(&self) -> native_pkcs11_traits::Result<Vec<Box<dyn Certificate>>> {
+        trace!("find_all_certificates");
+        Ok(vec![])
+    }
+
+    fn find_private_key(
+        &self,
+        _query: SearchOptions,
+    ) -> native_pkcs11_traits::Result<Option<Arc<dyn PrivateKey>>> {
+        trace!("find_private_key: {:?}", _query);
+        Ok(None)
+    }
+
+    fn find_public_key(
+        &self,
+        query: SearchOptions,
+    ) -> native_pkcs11_traits::Result<Option<Arc<dyn PublicKey>>> {
+        trace!("find_public_key: {:?}", query);
+        Ok(None)
+    }
+
+    fn find_all_private_keys(&self) -> native_pkcs11_traits::Result<Vec<Arc<dyn PrivateKey>>> {
+        trace!("find_all_private_keys");
+        Ok(vec![])
+    }
+
+    fn find_all_public_keys(&self) -> native_pkcs11_traits::Result<Vec<Arc<dyn PublicKey>>> {
+        trace!("find_all_public_keys");
+        Ok(vec![])
+    }
+
+    fn find_data_object(
+        &self,
+        query: SearchOptions,
+    ) -> native_pkcs11_traits::Result<Option<Arc<dyn DataObject>>> {
+        trace!("find_data_object: {:?}", query);
+        Ok(None)
+    }
+
+    fn find_all_data_objects(&self) -> native_pkcs11_traits::Result<Vec<Arc<dyn DataObject>>> {
+        trace!("find_all_data_objects");
+        let disk_encryption_tag = std::env::var("COSMIAN_PKCS11_DISK_ENCRYPTION_TAG")
+            .unwrap_or("disk-encryption".to_string());
+        let keys = get_pkcs11_keys(&self.kms_client, &[disk_encryption_tag])?;
+        Ok(keys
+            .into_iter()
+            .map(|dao| -> Arc<dyn DataObject> { Arc::new(dao) })
+            .collect())
```

Comment:
Maybe (I can't compile the KMS locally)
```suggestion
        Ok(keys
            .into_iter()
            .map(Arc::new)
            .collect())
```

Reply:
Unfortunately, the compiler rqauires the cast